### DISCUSSION TITLE
Re-import using search-id

### DISCUSSION THREAD

**Discussion by arsaboo** [beetbox/beets#4762] [Q&A]
I have the `spotify_album_id` for an album in the beets database. Can we specify the `search-id` in the database instead of typing it out during a re-import? The following does not work:
```
beet import "/data/music/Hindi Music/E/Ek Villain Returns (2022)/" --search-id $spotify_album_id
```
However, if I type out the `spotify_album_id`, it works fine:
```
beet import -m -I -t "/data/music/Hindi Music/E/Ek Villain Returns (2022)/" --search-id 6DCj9zSkRhSN77UQe3J4s1
```

**Comment by sampsyo** [comment]
Hmm; that's an interesting thought. Maybe the thing to do might be to make the importer do this automatically, with no extra flags? This is already true for MusicBrainz IDs, for instance: if a file already has an album MBID, we will attempt to search using that first, before we even bother using the textual metadata. Perhaps we can figure out a good way to give plugins the same "privilege"?

**Reply by arsaboo** [reply]
Are any of the plugins writing the identifiers (or any custom fields) to the file tags?

**Reply by sampsyo** [reply]
I don't think any currently do. But if this is a re-import (so the data is in flexible attributes in the database), it would be great if we could reuse that information.

**Reply by arsaboo** [reply]
Yes, that is what I am thinking. Do you have any thoughts on how to implement something like this (or any reference)? Given that there are multiple tag sources, I think this would be a great feature to have (I personally absolutely need this).

**Reply by sampsyo** [reply]
To summarize, the missing functionality here is to attempt to pull metadata _for the import process_ from the database instead of from disk, when we're doing a re-import. (An alternative for this particular use case would be to write the IDs to the file, FWIW, but maybe that's a separate issue.) It's undoubtedly going to be quite tricky, because the way the importer works is pretty subtle—but if you're interested, the relevant place could be `query_tasks`, where we first construct `ImportTask` objects based on querying the library:
https://github.com/beetbox/beets/blob/2bcf16b10c305a02489aec75e61c5d4728cd34b3/beets/importer.py#L1391

That would need to associate extra data from the database with each `ImportTask`. This would *only* work for `beet import -L` imports, however—which your examples above are not. For reimports triggered implicitly by importing paths that are already in the library, this actually seems considerably harder!

**Reply by arsaboo** [reply]
Wondering if the approach you suggested [here](https://github.com/beetbox/beets/discussions/4768#discussioncomment-5720858) (using formatted modify) can be used to import using `search-id`. I am thinking we can save the `id` in the `comment` (or any other field) and use that during the import process.

**Reply by sampsyo** [reply]
That's a good idea! You could "transfer" the information to a field that _is_ stored in on-disk tags with something like `beet modify comments=$spotify_album_id`. The only thing is that it would be really weird to hard-code beets to look for IDs in the `$comments` field… but maybe a hacked-up personal plugin could do it?

**Reply by arsaboo** [reply]
How about this....instead of using the `comments` field, we use the custom field, e.g., `beet modify spotify_album_id=$spotify_album_id`? Also, can you indicate which function should be modified to look for IDs in such custom fields....thanks!

**Reply by sampsyo** [reply]
Sure! Searching by "current ID" happens in `match_by_id`:
https://github.com/beetbox/beets/blob/2bcf16b10c305a02489aec75e61c5d4728cd34b3/beets/autotag/match.py#L304

That's hard-coded for MBIDs currently. You can see where it's invoked from `tag_album`, here:
https://github.com/beetbox/beets/blob/2bcf16b10c305a02489aec75e61c5d4728cd34b3/beets/autotag/match.py#L479

Basically, the logic says "see if there is an ID-based match and, if so, just use that. Otherwise, fall back to a full metadata-based search."

**Reply by arsaboo** [reply]
Ahh...that helps. I think it should be a relatively simple change to check for `spotify_album_id` here:
https://github.com/beetbox/beets/blob/2bcf16b10c305a02489aec75e61c5d4728cd34b3/beets/autotag/match.py#L309
and then call the respective `album_for_id` function. 

What would be the easiest way to extend the `match_by_id` function in a custom plugin? Basically, any way to supersede the default `match_by_id` function?

Also, would you be open to a PR for something like this? I will have to play with it a little to see the easiest way forward.

**Reply by sampsyo** [reply]
Good question! Yeah, happy to experiment even in a rough-draft PR.

The best way might be to emit a plugin event right at the top of `match_by_id`? If that plugin event returns anything, perhaps it should bypass the rest of the function.

**Comment by govynnus** [comment]
In reference to https://github.com/beetbox/beets/discussions/4784#discussioncomment-5850915, this might work?
```
beet list -a -f '"$path" $spotify_album_id' | xargs -o -n 2 sh -c 'beet import "$0" --search-id $1'
```

But reusing database info for reimports automatically could be nice :)

**Reply by arsaboo** [reply]
This is pretty close to perfect...the only issue is ` $spotify_album_id` is an item-level field. Thus, `beet list -a -f '"$path" $spotify_album_id'` only returns 
```
"/data/music/#/1947 Earth (1998)" $spotify_album_id
"/data/music/#/2.0 (2018)" $spotify_album_id
"/data/music/#/99 Songs (2020)" $spotify_album_id
```

**Reply by govynnus** [reply]
Ah, probably easiest is to set a flexible album level field first, like in https://github.com/beetbox/beets/discussions/4784#discussioncomment-5850469
```
beet ls -f '$album_id $spotify_album_id' album_id+ | uniq | xargs -o -n 2 sh -c 'beet modify -a id:$0 spotify_album_id=$1'
```
Bit hacky, but could be fine if you don't need to re-import regularly.

**Reply by arsaboo** [reply]
Thanks a ton @govynnus....this works:

```
beet ls -a -f '"$path" $alb_spotify_album_id' album_id+ data_source:"Spotify" | uniq |  xargs -o -n 2 sh -c 'beet import -m -I -t "$0" --search-id $1'
```
The inline field is:
```yaml
album_fields:
    alb_spotify_album_id: |
        for item in items:
            try:
                id = item.spotify_album_id
            except:
                id = None
        return id
```

Please suggest improvements, if any. Thanks!

**Reply by govynnus** [reply]
Ah cool, inline fields look nice but I've never used them :)
I wonder if you need the sorting (`album_id+`) and `unique`?

**Reply by arsaboo** [reply]
You are right, those are probably not needed.

**Reply by arsaboo** [reply]
For debugging purposes, it may be a good idea to add `set -x`...something like:
```
beet ls -a -f '"$path" $alb_spotify_album_id' |  xargs -o -n 2 sh -c 'set -x; beet import -m -I -t "$0" --search-id $1'
```

**Reply by JOJ0** [reply]
You could shorten the inline field by using .get method and falling back to None (the default of the get method when no value is found is None):


```
    alb_spotify_album_id: |
        for item in items:
            id = item.get('spotify_album_id')
        return id
```
